blog/2023-09-25-constructs-methods.md

```diff
@@ -0,0 +1,247 @@
+
+Today, Winglang uses **constructs**, a foundational framework that underpins several infrastructure-as-code libraries such as [AWS CDK], [CDKTF], and [cdk8s], to allow cloud components to be expressed and composed together into higher level abstractions.
+
+In the Wing application below, the classes named `Bucket` and `Function` are both constructs:
+
+```js
+bring cloud;
+
+let bucket = new cloud.Bucket();
+new cloud.Function(inflight () => {
+    bucket.put("Hello", "world");
+});
+```
+
+Each class in Winglang's standard library representing a cloud component is a construct.
+When an application is compiled, a construct is converted into one or more pieces of declarative [IaC] configuration (such as Terraform or CloudFormation).
+
+One superpower of constructs is that their properties can be configured after initialization, through methods.
+For example, environment variables can be added to a serverless function during or after initialization:
+
+```js
+let fn = new cloud.Function(
+  inflight () => { /* code */ },
+  env: { DB_HOST: "af43b12" }
+);
+
+// ... some more code
+
+fn.addEnvironment("DB_NAME", "orders");
+```
+
+However, this extra expressiveness introduces some challenges once we try to compose constructs together.
+In this blog post I'll highlight some of these challenges, and explain several of the best practices for designing APIs that avoid these pitfalls.
+
+[IaC]: https://en.wikipedia.org/wiki/Infrastructure_as_code
+[AWS CDK]: https://github.com/aws/aws-cdk
+[CDKTF]: https://github.com/hashicorp/terraform-cdk
+[cdk8s]: https://github.com/cdk8s-team/cdk8s
+
+## What is `constructs`?
+
+First, let's familiarize ourselves with `constructs` to get an idea of how it works.
+
+`constructs` at its core is a JavaScript library that provides a set of APIs for organizing classes into trees.
+A construct is created in JavaScript by writing a class that extends the `Construct` class, with a signature of `(scope, id, props)`.
+Constructs are always created in the scope of another construct[^1] and must always have an identifier which must be unique within the scope it’s created.
+A construct's identifier is used to generate a unique names for every cloud component.
+
+[^1]: An exception is the "root" construct, often named `App` or something similar.
+
+In this running example, I'll create a made-up construct framework for modeling gardens.
+Let's imagine our garden framework produces a `garden.json` file that declaratively specifies all of the flowers in our garden in a flat list.
+
+```js
+// garden.js
+const { Construct } = require("constructs");
+const { writeFileSync } = require("node:fs");
+
+// --- garden framework classes ---
+
+class Flower extends Construct {
+  constructor(scope, id, props) {
+    super(scope, id);
+    this.kind = props.kind;
+    this.color = props.color;
+  }
+
+  synth() {
```

Comment:
Traditionally we use something like `toJson()` at the node level and `synth()` only at the app level.